lib/ProductOpener/Display.pm

```diff
@@ -10332,6 +10272,285 @@ CSS
+			and ($product_ref->{nutriments}{$nid . "_modifier"} eq '-'))
+			or ((defined $product_ref->{nutriments}{$nid . "_prepared_modifier"})
+			and ($product_ref->{nutriments}{$nid . "_prepared_modifier"} eq '-'))
+	);
+
+	# Logic to show important or non-optional nutrients
+	if (($nutriment !~ /-$/) or $is_nutrient_with_value) {
+		return 1;
+	}
+
+	# Hide non-important nutrients when no value is present
+	if ((($nutriment !~ /^!/) or ($product_ref->{id} eq 'search')) and not($is_nutrient_with_value)) {
+		return 0;
+	}
+
+	# Special handling for nutrition scores (optional, based on request_ref)
+	if ($nid =~ /^nutrition-score-(.*)$/) {
+		# Always show FR score and Nutri-Score
+		if (($request_ref->{cc} ne $1) and (not($1 eq 'fr'))) {
+			return 0;
+		}
+
+		# Experimental: hide nutrition scores for now
+		return 0;
+	}
+
+	return 1;
+}
+
+=head2 get_nutrient_level ( $nutriment )
+
+Determines the level of a nutrient based on the number of leading dashes.
+
+The level is used to establish whether a nutrient is a top-level nutrient or a sub-nutrient (e.g., fat -> saturated fat -> omega-3).
+
+=head3 Arguments
+
+=head4 Nutrient $nutriment
+
+The nutrient identifier string, which may contain leading dashes to denote levels.
+
+=head3 Return values
+
+Returns an integer indicating the nutrient's level.
+- 0 for main nutrients
+- 1 for sub-nutrients
+- 2 for sub-sub-nutrients, etc.
+
+=cut
+
+sub get_nutrient_level {
+	my ($nutriment) = @_;
+	if ($nutriment =~ /^!?--/) {
+		return 2;    # Sub-sub-nutrient (level 2)
+	}
+	elsif ($nutriment =~ /^!?-/) {
+		return 1;    # Sub-nutrient (level 1)
+	}
+	return 0;    # Top-level nutrient (level 0)
+}
+
+=head2 add_parent_if_missing ( $nutriment, $level, $template_data_ref, $added_nutrients_ref, $nutriments )
+
+Ensures that a parent nutrient (e.g., polyol for isomalt) is added to the nutrition table before adding the current nutrient.
+
+This function looks for the parent nutrient in the list of nutrients and adds it to the template if it's missing.
+
+=head3 Arguments
+
+=head4 Nutrient $nutriment
+
+The current nutrient being processed (e.g., isomalt).
+
+=head4 Level $level
+
+The level of the current nutrient, which helps determine the parent nutrient (level - 1).
+
+=head4 Template data reference $template_data_ref
+
+A reference to the data structure that holds the nutrition table rows.
+
+=head4 Added nutrients reference $added_nutrients_ref
+
+A hash reference tracking nutrients that have already been added to avoid duplication.
+
+=head4 Nutriments array reference $nutriments
+
+A reference to the array of all nutrients to locate the parent nutrient.
+
+=head3 Return values
+
+None. The function modifies the template data in-place.
+
+=cut
+
+sub add_parent_if_missing {
+	my ($nutriment, $level, $template_data_ref, $nutriments, $product_ref) = @_;
+
+	# top nutrients don't have parents
+	return if $level == 0;
+
+	# Locate the parent nutrient by checking the previous nutrients in @nutriments
```

Comment:
It feels very cumbersome to do this at runtime, going backward to get the parent etc.

A simpler approach would be to create a hash of all nutrients at initialization time, that would store the level and parent of all nutrients. That way at runtime we don't need to recompute it, and the code will be much shorter.

Search for "# Compute the list of nutriments that are not shown by default so that they can be suggested" in Food.pm, we already do something similar.